src/execution_plan/optimizations/reduce_traversal.c

```diff
@@ -0,0 +1,109 @@
+ * with expand-into op. */
+void reduceTraversal(ExecutionPlan *plan, AST *ast) {
+    OPType t = OPType_CONDITIONAL_TRAVERSE | OPType_CONDITIONAL_VAR_LEN_TRAVERSE;
+    OpBase **traversals = ExecutionPlan_LocateOps(plan->root, t);
+    uint traversals_count = array_len(traversals);
+    
+    for(uint i = 0; i < traversals_count; i++) {
+        OpBase *op = traversals[i];
+        AlgebraicExpression *ae;
+
+        if(op->type == OPType_CONDITIONAL_TRAVERSE) {
+            CondTraverse *traverse = (CondTraverse*)op;
+            ae = traverse->algebraic_expression;
+        } else if(op->type == OPType_CONDITIONAL_VAR_LEN_TRAVERSE) {
+            CondVarLenTraverse *traverse = (CondVarLenTraverse*)op;
+            ae = traverse->ae;
+        } else {
+            assert(false);
+        }
+
+        /* If traverse src and dest nodes are the same,
+         * traverse acts as a filter which make sure the node is of a specific type
+         * e.g. MATCH (a:A)-[e:R]->(b:B) RETURN e 
+         * in this case there will be a traverse operation which will 
+         * filter our dest nodes (b) which aren't of type B. */
+        if(ae->src_node == ae->dest_node) continue;
+
+        /* Search to see if dest is already resolved */
+        const char *dest = ae->dest_node->alias;
+        if(!_entity_resolved(op->children[0], dest)) continue;
+
+        /* Both src and dest are already known
+         * perform expand into instaed of traverse. */
+        if(op->type == OPType_CONDITIONAL_TRAVERSE) {
+            CondTraverse *traverse = (CondTraverse*)op;
+            OpBase *expand_into = NewExpandIntoOp(traverse->algebraic_expression, ast);                                                  
+
+            // Set traverse algebraic_expression to NULL to avoid early free.
+            traverse->algebraic_expression = NULL;
+            ExecutionPlan_ReplaceOp(plan, (OpBase*)traverse, expand_into);
+            OpBase_Free((OpBase*)traverse);
+        } else {
+            CondVarLenTraverse *traverse = (CondVarLenTraverse*)op;
+            CondVarLenTraverseOp_ExpandInto(traverse);
+            /* Conditional variable length traversal do not perform
+             * label filtering by matrix matrix multiplication
+             * it introduces conditional traverse operation in order 
+             * to perform label filtering, but in case a node is already 
+             * resolved this filtering is redundent and should be removed. */
+            OpBase *t;
+            if(ae->src_node->label) {
+                t = op->children[0];
+                if(t->type == OPType_CONDITIONAL_TRAVERSE) {
+                    _removeRedundentTraversal(plan, (CondTraverse*)t);
+                }
+            }
+            if(ae->dest_node->label) {
+                t = op->parent;
+                if(t->type == OPType_CONDITIONAL_TRAVERSE) {
+                    _removeRedundentTraversal(plan, (CondTraverse*)t);
```

Comment:
Memory leak on ops freed this way (line 95 too), but since those ops are in the `traversals` array, this might have to be handled with care.

Reply:
👍